src/main/java/io/gravitee/policy/generatehttpsignature/GenerateHttpSignaturePolicy.java

```diff
@@ -0,0 +1,167 @@
+            logger.warn(checkHeadersErrorMessage);
+            chain.failWith(PolicyResult.failure(HTTP_SIGNATURE_IMPOSSIBLE_GENERATION, 400, checkHeadersErrorMessage));
+            return;
+        }
+
+        final Signature signatureFromConfiguration = buildSignatureFromConfiguration(context, request, configuredHeaders);
+
+        final Signer signer = buildSigner(context, signatureFromConfiguration);
+        final Signature signature;
+
+        try {
+            signature = signer.sign(request.method().name().toLowerCase(), request.path(), requestHeaders.toSingleValueMap());
+        } catch (IOException e) {
+            final String errorMessage = String.format("%s %s", ERROR_MESSAGE, e.getMessage());
+            logger.warn(errorMessage);
+            request.metrics().setMessage(e.getMessage());
+            chain.failWith(PolicyResult.failure(HTTP_SIGNATURE_IMPOSSIBLE_GENERATION, 400, errorMessage));
+            return;
+        }
+
+        setSignatureHeader(request, signature);
+
+        chain.doNext(request, response);
+    }
+
+    /**
+     * Validate the request by verifying it has the headers mentioned in policy's configuration.
+     * @param headers the incoming request's headers.
+     * @param configuredHeaders the headers from policy's configuration.
+     * @return the error message if headers are not valid, null if it's ok
+     *
+     */
+    String checkHeaders(HttpHeaders headers, List<String> configuredHeaders, PolicyChain chain) {
+        if (!isEmpty(configuredHeaders)) {
+            if (!headers.containsAllKeys(configuredHeaders)) {
+                final ArrayList<String> missingHeaders = new ArrayList<>(configuredHeaders);
+                missingHeaders.removeAll(headers.keySet());
+
+                return String.format(
+                    "%s those headers are missing %s",
+                    ERROR_MESSAGE,
+                    missingHeaders.stream().collect(Collectors.joining(", ", "[", "]"))
+                );
+            }
+        } else if (!headers.containsKey(HttpHeaders.DATE)) {
+            return String.format("%s 'Date' header is missing", ERROR_MESSAGE);
+        }
+
+        return null;
+    }
+
+    /**
+     * Build signature from policy configuration.
+     * @param context the API execution context
+     * @param configuredHeaders the headers to add to signature
+     * @return the signature
+     */
+    Signature buildSignatureFromConfiguration(ExecutionContext context, Request request, List<String> configuredHeaders) {
+        final String keyId = context.getTemplateEngine().getValue(configuration.getKeyId(), String.class);
+        return new Signature(
+            keyId,
+            null,
+            configuration.getAlgorithm().getAlg(),
+            null,
+            null,
+            configuredHeaders,
+            configuration.getValidityDuration() * 1000,
+            getRequestTimestamp(request),
+            null
```

Comment:
I wonder if the expirationTime should not be set here ?

Reply:
with the library we are using we can use it like that